registry/reference.go

```diff
@@ -56,33 +56,53 @@ type Reference struct {
 	Reference string
 }
 
-// ParseReference parses a string into a artifact reference.
-// If the reference contains both the tag and the digest, the tag will be
-// dropped.
+// ParseReference parses a string into an artifact reference.
+// If the reference contains both the tag and the digest, the tag will be dropped.
 // Digest is recognized only if the corresponding algorithm is available.
-func ParseReference(raw string) (Reference, error) {
-	parts := strings.SplitN(raw, "/", 2)
+//
+// The parameter `image` is an opaque polymorphic string, so it's worth layout out the 4 valid forms that it can take.
+// This parameter can be parsed as follows; *lowercase* labels here imply opaque values (composed of other components),
+// and *uppercase* labels imply resolute values (can not be expanded any further).
+//
+//		<---------------- image ---------------------------------------------------> | Decode `image`
+//		<=== REGISTRY ===> / <------------------ path -----------------------------> |  - Decode `path``
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> * <---------- reference ---------> |    - Decode `reference`
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> @ <=================== DIGEST ===> |      - Valid Form A
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ===> @ <=== DIGEST ===> |      - Valid Form B
+//		<=== REGISTRY ===> / <=== REPOSITORY ===> : <=== TAG ======================> |      - Valid Form C
+//		<=== REGISTRY ===> / <=== REPOSITORY ======================================> |    - Valid Form D
+//
+//	Note: DIGEST is itself, if present, of the form
+//		<=========== DIGEST ==========>
+//		<=== ALGO ===> : <=== HASH ===>
+//
+//	Note: Furthermore, in the case of Valid Form B, the TAG is dropped without any validation or consideration
+//		  whatsoever.
```

Comment:
Good idea, BNF is much more compact so I can defined the entire range of possibilities.  I might leave the section of the existing docstring though, since it helps to map the forms to the 4x valid forms.  Let me know what you think once I publish it.

Reply:
Oh, and the godoc looks reasonable to me, let me know if you would like a screenshot.